sklearn/preprocessing/tests/test_encoders.py

```diff
@@ -827,31 +829,37 @@ def test_encoders_has_categorical_tags(Encoder):
     assert 'categorical' in Encoder()._get_tags()['X_types']
 
 
-@pytest.mark.parametrize('input_dtype', ['O', 'U'])
-@pytest.mark.parametrize('category_dtype', ['O', 'U'])
+@pytest.mark.parametrize('input_dtype', ['O', 'U', 'S'])
+@pytest.mark.parametrize('category_dtype', ['O', 'U', 'S'])
 @pytest.mark.parametrize('array_type', ['list', 'array', 'dataframe'])
-def test_encoders_unicode_categories(input_dtype, category_dtype, array_type):
-    """Check that encoding work with string and object dtypes.
+def test_encoders_string_categories(input_dtype, category_dtype, array_type):
+    """Check that encoding work with object, unicode, and byte string dtypes.
     Non-regression test for:
     https://github.com/scikit-learn/scikit-learn/issues/15616
     https://github.com/scikit-learn/scikit-learn/issues/15726
     """
 
+    # Expect incompatible dtypes to raise an exception on fit(X)
+    may_pass = (category_dtype, input_dtype) not in {('S', 'O')}
+    context = raises(ValueError, match="unknown categories", may_pass=may_pass)
+
     X = np.array([['b'], ['a']], dtype=input_dtype)
+    X_test = _convert_container([['a'], ['a'], ['b'], ['a']],
+                                array_type, dtype=input_dtype)
```

Comment:
@thomasjpfan I just want to confirm that below is OK, without *np.array*:
```python
    if input_dtype == 'S':
        X_test = [[b'a'], [b'a'], [b'b'], [b'a']]  # S requires bytes
    else:
        X_test = [['a'], ['a'], ['b'], ['a']]
    X_test = _convert_container(X_test, array_type)
```

Reply:
That doesn't work for `array_type="dataframe"`.

Reply:
@thomasjpfan It works for dataframe in pandas v1.2.3, so I just want to make sure this is a concern about earlier versions.

Reply:
(pandas 1.2.0) The bytes dtype is not carried over for dataframes:

```py
a_bytes = np.array([[b"1"], [b"3"]])
_convert_container(a_bytes, "dataframe").dtypes
# 0    |S1
# dtype: object
```

```py
a_list_bytes = [[b"1"], [b"3"]]
_convert_container(a_list_bytes, "dataframe").dtypes
# 0    object
# dtype: object
```

Reply:
I see. Still, the test passes for *category_dtype=S* with or without the ndarray (encoding of *X_test* works correctly for dataframe dtype of O or S). I also don't think it's OK to make *X_test* an ndarray since then *array_type="list"* will produce
```python
>>> _convert_container(np.array([[b'a'], [b'b']]), "list")
[array([b'a'], dtype='|S1'), array([b'b'], dtype='|S1')]
```
which is not what the original test was testing.

A compromise might be to add a case that uses ndarray for *array_type="dataframe"*:
```python
    if input_dtype == 'S':
        # bytes require th input to be consistent type
        X_test = [[b'a'], [b'a'], [b'b'], [b'a']]
        if array_type == 'dataframe':
            X_test = np.array(X_test)
    else:
        X_test = [['a'], ['a'], ['b'], ['a']] 
```

Reply:
Wow. At this point, I think it would be cleaner to add `dtype` to `_convert_container` like you originally had in mind.

Thank you for your patience on this. The encoders are always a little more complicated than they look.